File: LeetCode Problems/901 to 1000/947. Most Stones Removed with Same Row or Column/947. Most Stones Removed with Same Row or Column.cpp
class Solution {
    unordered_map<int, int> f;
    int res = 0;

public:
    int removeStones(vector<vector<int>> &stones) {
        for (int i = 0; i < stones.size(); i++)
            Union(stones[i][0], ~stones[i][1]);
        return stones.size() - res;
    }
    int find(int x) {
        if (!f.count(x))
            f[x] = x, res++;
        if (x != f[x])
            f[x] = find(f[x]);
        return f[x];
    }
    void Union(int x, int y) {
        x = find(x), y = find(y);
        if (x != y)
            f[x] = y, res--;
    }
};

File: LeetCode Problems/901 to 1000/947. Most Stones Removed with Same Row or Column/947. Most Stones Removed with Same Row or Column.java
class Solution {
    private Map<Integer, Integer> f = new HashMap<>();
    private int res = 0;
    public int removeStones(int[][] stones) {
        for (int i = 0; i < stones.length; i++) {
            union(stones[i][0], ~stones[i][1]);
        }
        return stones.length - res;
    }
    private void union(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) {
            f.put(x, y);
            res--;
        }
    }
    private int find(int x) {
        if (f.putIfAbsent(x, x) == null) res++;
        if (x != f.get(x)) f.put(x, find(f.get(x)));
        return f.get(x);
    }
}
